src/socket/communicate.rs

```diff
@@ -120,12 +119,15 @@ pub mod listener {
         accept_timeout: Timeout,
     }
 
-    /// A thread spawned by the accept method.
-    pub struct Thread {
-        /// The kind of handler that is run by this thread (e.g. a ping event)
-        pub message_type: CommunicationType,
-        /// Handle to join the thread
-        pub handle: thread::JoinHandle<()>,
+    /// Standing connection to the client.
+    /// The `communication_type` determines which handler the user has to call in `Handlers`,
+    /// but the missing dependent types makes it hard to express;
+    /// Handlers should probably be a trait instead.
```

Comment:
Couldn't you make `CommunicationType` an `Enum` which contains either `Ping` or `StreamEvents`? You could then make `Handler` contain a `perform_communication` thingy which takes a socket…

Reply:
The idea is that you have a message in the beginning, which tells the server what kind of event the sockets should talk, and then after you get a `<R, W>` where R and W are fixed to the message types.

Reply:
Basically the R and W types are dependent on which kind of message the client sends first. Which you can do manually like here, but you could possibly also create a trait that encodes that somehow.

Reply:
Couldn't you attach the getting the ReadWriter part to the CommunicationType rather than the Handlers struct as the former actually is related to the first message?

Reply:
I don’t think so, but I might have overcomplicated this‽

Reply:
I'm not certain, I guess I would have to try out my idea.